components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/BookmarksStorageSuggestionProvider.kt

```diff
@@ -0,0 +1,49 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.awesomebar.provider
+
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.concept.awesomebar.AwesomeBar
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarksStorage
+import mozilla.components.feature.awesomebar.internal.loadLambda
+import mozilla.components.feature.session.SessionUseCases
+import java.util.UUID
+
+private const val BOOKMARKS_SUGGESTION_LIMIT = 20
+
+/**
+ * A [AwesomeBar.SuggestionProvider] implementation that provides suggestions based on the bookmarks
+ * stored in the [BookmarksStorage].
+ */
+class BookmarksStorageSuggestionProvider(
+    private val bookmarksStorage: BookmarksStorage,
+    private val loadUrlUseCase: SessionUseCases.LoadUrlUseCase,
+    private val icons: BrowserIcons? = null
+) : AwesomeBar.SuggestionProvider {
+
+    override val id: String = UUID.randomUUID().toString()
+
+    override suspend fun onInputChanged(text: String): List<AwesomeBar.Suggestion> {
+        if (text.isEmpty()) {
+            return emptyList()
+        }
+
+        val suggestions = bookmarksStorage.searchBookmarks(text, BOOKMARKS_SUGGESTION_LIMIT)
+        return suggestions.distinctBy { it.url }.into()
+    }
+
+    private fun List<BookmarkNode>.into(): List<AwesomeBar.Suggestion> {
+        return this.map {
+            AwesomeBar.Suggestion(
+                provider = this@BookmarksStorageSuggestionProvider,
+                icon = icons.loadLambda(it.url!!),
```

Comment:
The underlying API doesn't guarantee that there's a `url`, which is why you're casting here. From their docs: "@return A list of bookmarks where either the URL or the title contain a word (e.g. space separated item) from the query."

This reads to me like you may get back a folder. and you probably don't want to throw if the `url` isn't there. This is something you can double-check above at the `onInputChanged` level.

Reply:
I added two filters one here and one in `onInputChanged` to make sure that the URLs aren't `null`, but I don't think map is smart enough to not let me have the `!!` in both places...